### ISSUE TITLE
Support ranges

### ISSUE THREAD

**Issue by GoogleCodeExporter** [nodatime/nodatime#2]
```
Several applications need the idea of a range - I suspect the most common 
case is local dates.

It would be nice if we could use:

    LocalDateRange range = new LocalDateRange(start, end);
    foreach (LocalDate date in range)
    {
    }

We can use the code in MiscUtil (http://pobox.com/~skeet/csharp/miscutil) as 
a starting point for this. We may want to consider whether to "go generic" or 
just include the ranges for the most common cases.
```

Original issue reported on code.google.com by `jonathan.skeet` on 4 Dec 2009 at 8:09

**Comment by jskeet** [comment]
This is implemented as `DateInterval`. Leaving this open to review the API.

**Comment by lundmikkel** [comment]
@jskeet Date intervals are normally closed intervals, i.e. both endpoints should be included. @mj1856 should be able to confirm.

**Comment by jskeet** [comment]
@lundmikkel: I think it can be useful to have the *ability* to treat them as half-closed, e.g. to make it clear that a sequence of intervals is contiguous. (e.g. "1st of January (inclusive) to 1st February (exclusive), then 1st February (inclusive) to 1st March (exclusive)") The current `DateInterval` allows both... do you think that causes any problems? It does mean users need to be aware that it _can_ be used both ways, admittedly.

**Comment by lundmikkel** [comment]
I find that people often get confused about interval endpoint inclusion. For certain interval types, only/mainly one type of endpoint inclusion makes sense. E.g. time intervals are half-open (as with `Interval`), date intervals are closed. That doesn't necessarily mean that that is the only way to consider them, but keeping to a "standard" would be nice.

I have previously done an [extensive interval collection library](https://github.com/lundmikkel/c5intervals), and we found that doing endpoint inclusion explicit was the best help for users. It also makes algorithms, e.g. for overlap or containment, easier since they can be considered more generally.

To get back to `DateInterval`, I think calling it `Inclusive` is asking for trouble. It isn't clear what kind of inclusion it is, unless you read the documentation, which states that it only is the last endpoint. My suggestion is to name it explicitly, e.g. 'EndInclusive', or only allow one type of interval, e.g. closed.

Another possible approach is the one we took, where the interval is generic and allows for all types of endpoint inclusion, but I haven't looked enough at `DateInterval` to say if that makes sense.

**Comment by jskeet** [comment]
I would definitely be happy with `EndInclusive` instead of `Inclusive`. But I think there are sufficient situations where you *do* want an exclusive end point for a date range that it's worth keeping that available. (I don't think it would make sense for a regular `Interval` to be customizable like that though.)

**Comment by mj1856** [comment]
Well, perhaps I'm opinionated on this, but I really think `2017-01-01` - `2017-01-04` can *only* be interpreted as "four days", while `2017-01-01T00:00:00Z` - `2017-01-04T00:00:00Z` can *only* be interpreted as "72 hours" (or "3 standard days").

I can't really think of any exceptions to this, other than human confusion around whether or not the time is a part of the data in question - but we've eliminated that in NodaTime by having separate types for `LocalDate` and `Instant` (and others).

**Comment by jskeet** [comment]
I can't remember the discussion I had with @malcolmr about using different types for this, but possible compromise solution - 3 types:

- `IDateInterval` interface, with the same public members as the current `DateInterval`
- `DateInterval` class which is *always* end-inclusive
- `EndExclusiveDateInterval` which is *always* end-exclusive

The current `ContainedDatesDateIntervalEqualityComparer` could implement `IEqualityComparer<IDateInterval>`; `NormalizingDateIntervalEqualityComparer` would only be relevant to `EndExclusiveDateInterval` as `DateInterval` could never be empty.

Basically, those who never need to deal with an end-exclusive date interval could just use `DateInterval` (with the bias towards that being shown in the name) but for the uses as in my earlier comment, we could use `EndExclusiveDateInterval`.

Thoughts?

**Comment by lundmikkel** [comment]
I would seriously **only** have a `DateInterval` with both endpoints included. I don't see any real benefits from allowing both endpoint inclusion types. Quite the contrary actually: I think it only confuses the user even more.

I not sure about your example, @jskeet. If you wanted to create a half-open interval, then a factory method could be the solution. If it was for readability, then maybe the ToString/DebuggerDisplay could print month names instead of the specific dates. For instance (for some locale):

- [2017-01-01;2017-01-04] could print something like "January 1st, 2017 - January 4th, 2017"
- [2017-01-01;2017-01-31] could print something like "January, 2017"
- [2017-01-01;2017-12-31] could print something like "2017"

**Comment by jskeet** [comment]
@lundmikkel: That would assume that every use case for an interval that happens to cover every day of a month is the same. Maybe I've got a set of intervals that is *meant* to be February 1st to February 28th each year, exactly 28 days... it would be odd for that to show differently in leap years to non-leap-years.

However, the fact that *everyone* else seems to think the use case I'm concerned with is irrelevant does make me want to give in. @malcolmr - apologies for the fact that we'll be wasting a lot of the comparer work you did, if we do indeed make these always-inclusive.

**Comment by malcolmr** [comment]
(as noted in #671, I'd be happier if the inclusivity of the range was a property of the type rather than the instance anyway: it seems less prone to mistakes.)

**Comment by jskeet** [comment]
@malcolmr: Yup, hence my alternative proposal a few comments ago - but it sounds like Mikkel and Matt are pretty sure that the speculated `EndInclusiveDateInterval` type just shouldn't exist. I suppose we can always add that (and the interval) later...
